compiler/class-assumptions.cpp

```diff
@@ -114,6 +114,46 @@ ClassPtr Assumption::extract_instance_from_type_hint(const TypeHint *a) {
     a = as_optional->inner;
   }
 
+  if (const auto *as_pipe = a->try_as<TypeHintPipe>()) {
+    kphp_assert(as_pipe->items.size() >= 2);
+    // to avoid the redundant work, check whether all union variants are classes
+    // and that they have some parent or interface (otherwise they would never
+    // have a common base)
+    for (const auto *item : as_pipe->items) {
+      const auto *as_instance = item->try_as<TypeHintInstance>();
```

Comment:
may be it's worth to check TypeHintOptional case as well, so the following example would work

```php
<?php

class A {
  public $a = 1;
}

class A1 extends A {
  public $a1 = 1;
}

class A2 extends A {
  public $a2 = 1;
}

/** @param A1|A2|null $x */
function foo($x) {
  if ($x instanceof A1) {
    var_dump($x->a);
  }
}

foo(new A1);
foo(new A2);
```

Reply:
`A1|A2|null` is a pipe. TypeHintOptional is T|null (a single T, not many)

Reply:
I made `A1|A2|null` work by skipping `null` types during the common base type searching.
It doesn't make code much more complicated but looks quite useful. :+1: